exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java

```diff
@@ -130,562 +145,248 @@ public IterOutcome innerNext() {
-      // get the renamed column names
-      SchemaPath outputPath = SchemaPath.getSimplePath(outputFields.get(index).getPath());
+    for(VectorWrapper<?> vw : inputBatch) {
+      ValueVector vvIn = vw.getValueVector();
+      ValueVector vvOut = container.getValueVector(index).getValueVector();
 
       final ErrorCollector collector = new ErrorCollectorImpl();
       // According to input data names, Minortypes, Datamodes, choose to
       // transfer directly,
       // rename columns or
       // cast data types (Minortype or DataMode)
-      if (hasSameTypeAndMode(outputFields.get(index), vw.getValueVector().getField())) {
+      if (hasSameTypeAndMode(container.getSchema().getColumn(index), vvIn.getField())
+          && vvIn.getField().getType().getMinorType() != TypeProtos.MinorType.MAP // Per DRILL-5521, existing bug for map transfer
+          ) {
         // Transfer column
-
-        MajorType outputFieldType = outputFields.get(index).getType();
-        MaterializedField outputField = MaterializedField.create(outputPath.getAsUnescapedPath(), outputFieldType);
-
-        /*
-          todo: Fix if condition when DRILL-4824 is merged
-          If condition should be changed to:
-          `if (outputFields.get(index).getPath().equals(inputPath.getAsUnescapedPath())) {`
-          DRILL-5419 has changed condition to correct one but this caused regression (DRILL-5521).
-          Root cause is missing indication of child column in map types when it is null.
-          DRILL-4824 is re-working json reader implementation, including map types and will fix this problem.
-          Reverting condition to previous one to avoid regression till DRILL-4824 is merged.
-          Unit test - TestJsonReader.testKvgenWithUnionAll().
-         */
-        if (outputFields.get(index).getPath().equals(inputPath)) {
-          ValueVector vvOut = container.addOrGet(outputField);
-          TransferPair tp = vvIn.makeTransferPair(vvOut);
-          transfers.add(tp);
+        TransferPair tp = vvIn.makeTransferPair(vvOut);
+        transfers.add(tp);
         // Copy data in order to rename the column
-        } else {
-          final LogicalExpression expr = ExpressionTreeMaterializer.materialize(inputPath, current, collector, context.getFunctionRegistry() );
-          if (collector.hasErrors()) {
-            throw new SchemaChangeException(String.format("Failure while trying to materialize incoming schema.  Errors:\n %s.", collector.toErrorString()));
-          }
-
-          ValueVector vv = container.addOrGet(outputField, callBack);
-          allocationVectors.add(vv);
-          TypedFieldId fid = container.getValueVectorId(SchemaPath.getSimplePath(outputField.getPath()));
-          ValueVectorWriteExpression write = new ValueVectorWriteExpression(fid, expr, true);
-          cg.addExpr(write);
-        }
-      // Cast is necessary
+      } else if (vvIn.getField().getType().getMinorType() == TypeProtos.MinorType.NULL) {
```

Comment:
Why do we need to copy data to *rename* a column? The underlying buffer is identical regardless of name? Can't rename be done with a transfer? Or, is the original comment simply wrong?

Reply:
The comment is for the `else` branch, and should be moved a couple lines down. For rename a column, I do not think we are doing a copy; transfer should be good enough.